pkg/virt-handler/cache/cache.go

```diff
@@ -59,6 +70,216 @@ type DomainWatcher struct {
 	watchdogTimeout          int
 	recorder                 record.EventRecorder
 	vmiStore                 cache.Store
+
+	watchDogLock        sync.Mutex
+	unresponsiveSockets map[string]int64
+}
+
+type ghostRecord struct {
+	Name       string    `json:"name"`
+	Namespace  string    `json:"namespace"`
+	SocketFile string    `json:"socketFile"`
+	UID        types.UID `json:"uid"`
+}
+
+var ghostRecordGlobalCache map[string]ghostRecord
+var ghostRecordGlobalMutex sync.Mutex
+var ghostRecordDir string
+
+// this function is only used by unit tests
+func clearGhostRecordCache() {
+	ghostRecordGlobalMutex.Lock()
+	defer ghostRecordGlobalMutex.Unlock()
+	ghostRecordGlobalCache = make(map[string]ghostRecord)
+}
+
+func InitializeGhostRecordCache(directoryPath string) error {
+	ghostRecordGlobalMutex.Lock()
+	defer ghostRecordGlobalMutex.Unlock()
+
+	ghostRecordGlobalCache = make(map[string]ghostRecord)
+	ghostRecordDir = directoryPath
+	err := os.MkdirAll(ghostRecordDir, 0755)
+	if err != nil {
+		return err
+	}
+
+	files, err := ioutil.ReadDir(ghostRecordDir)
+	if err != nil {
+		return err
+	}
+
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+		recordPath := filepath.Join(ghostRecordDir, file.Name())
+
+		fileBytes, err := ioutil.ReadFile(recordPath)
+		if err != nil {
+			return err
+		}
+
+		ghostRecord := ghostRecord{}
+		err = json.Unmarshal(fileBytes, &ghostRecord)
+		if err != nil {
+			return err
```

Comment:
Do we want to log the error and continue, or do we want to basically make virt-handler stop working when an invalid file is put into the directory? Do we need some kind of mini-transactions here, so that we can clean up, in case that virt-handler got killed while writing an entry? E.g. write to `.transaction` file and do a `rename` afterwards?

Reply:
Thinking about this more, I'm in favor of logging the error and not panicing virt-handler.

> Do we need some kind of mini-transactions here, so that we can clean up, in case that virt-handler got killed while writing an entry? E.g. write to .transaction file and do a rename afterwards?

wouldn't this only be an issue on a non journalling file system? I'm not sure this is really necesssary. I just don't like the idea that a rogue file in that directory can prevent virt-handler from starting. Someone debugging could write something in that directory and hose the node the way it is now.